File: dogfood/snippets/Python/merge_sort.py
def merge(x, y):
    if x == []: return y
    if y == []: return x
    return [x[0]] + merge(x[1:], y) if x[0] < y[0] else [y[0]] + merge(x, y[1:])


def sort(a, n):
    m = n // 2
    return a if n <= 1 else merge(sort(a[:m], m), sort(a[m:], n - m))


a = [
    17,
    12,
    19,
    10,
    15,
    8,
    2,
    16,
    5,
    14,
    18,
    6,
    7,
    13,
    4,
    1,
    9,
    11,
    3,
    20,
]
print(" ".join(str(n) for n in sort(a, len(a))))

File: dogfood/snippets/Javascript/merge_sort.js
function mergeSort(v) {
    if (v.length <= 1) {
        return v;
    }

    let m = Math.floor(v.length / 2);
    let l = mergeSort(v.slice(0, m));
    let r = mergeSort(v.slice(m));
    return merge(l, r);

    function merge(a, b) {
        let i = 0, j = 0;
        let n = a.length + b.length;
        let c = [];
        while (c.length < n) {
            if (i < a.length && (j >= b.length || a[i] < b[j])) {
                c.push(a[i++]);
            } else {
                c.push(b[j++]);
            }
        }
        return c;
    }
}

console.log(mergeSort([2, 8, 3, 10, 13, 6, 11, 9, 19, 15, 5, 4, 12, 14, 20, 1, 17, 18, 16, 7]).join(" "));
